File: Chapters/Chapter_18/1D_Advection_Diffusion/create_rhs.m
%---------------------------------------------------------------------%
%This function computes the LGL grid and elements.
%Written by F.X. Giraldo on 10/2003
%           Department of Applied Mathematics
%           Naval Postgraduate School
%           Monterey, CA 93943-5216
%---------------------------------------------------------------------%
function rhs = create_rhs(qp,coord,nelem,ngl,nq,wnq,psi,dpsi,u)

%Initialize
rhs=zeros(ngl,nelem);
q_e=zeros(ngl,1);
x=zeros(ngl,1);

%Integrate Divergence of Flux
for e=1:nelem
   
   %Store Coordinates
   q_e(:)=qp(:,e);
   x(:)=coord(:,e);
   
   %Jacobians
   dx=x(ngl)-x(1);
   jac=dx/2;
   ksi_x=2/dx;
   
   %LGL Integration
   for l=1:nq
      wq=wnq(l)*jac;
            
      %Form Derivative
      q_k=0;
      q_ksi=0;
      for j=1:ngl
          q_k=q_k + psi(j,l)*q_e(j)*u;
          q_ksi=q_ksi + dpsi(j,l)*q_e(j);
      end
      q_x=q_ksi*ksi_x;
      
      %Form RHS
      for i=1:ngl
         dhdx_i=dpsi(i,l)*ksi_x;
%          rhs(i,e)=rhs(i,e) + wq*dhdx_i*q_k - 0*wq*visc*dhdx_i*q_x;
         rhs(i,e)=rhs(i,e) + wq*dhdx_i*q_k;

      end %i
      
   end %l
end %e

File: Projects/Project_02_1D_Wave/For_Instructors/julia/Global_Matrices_approach/create_rhs.jl
#---------------------------------------------------------------------#
#This function computes the global Mass and Differentiation Matrices.
#Written by F.X. Giraldo on April 19 2019
#           Department of Applied Mathematics
#           Naval Postgraduate School
#           Monterey; CA 93943-5216
#---------------------------------------------------------------------#
function create_rhs(q,u,Dhat,Fhat,intma,DFloat)

    #Get lengths of arrays
    (Np,nelem)=size(intma)
    Npoin=size(q)

    #store flux term
    f=q*u

    #build numerical flux: fstar
    fstar=zeros(DFloat,Npoin)
    for e=1:nelem
        L=intma[Np,e]
        ep1=e+1
        if (e == nelem)
            ep1=1
        end
        R=intma[1,ep1]
        flux=0.5*( f[L] + f[R] - u*(q[R] - q[L]) )
        fstar[L]=flux
        fstar[R]=flux
    end

    #Build RHS vector
    R=Dhat*f - Fhat*fstar

    #Return arrays
    return(R)
end
